app/app.go

```diff
@@ -0,0 +1,108 @@
+package app
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+
+	"github.com/grafana/grafana-app-sdk/resource"
+)
+
+var (
+	// ErrNotImplemented is an error that indicates that an App method is not implemented by the App implementation,
+	// or the App method is not implemented for the provided Kind.
+	// Typically, the App's ManifestData should indicate this as well.
+	ErrNotImplemented = errors.New("not implemented")
+)
+
+// ConversionRequest is a request to convert a Kind from one version to another
+type ConversionRequest struct {
+	SourceGVK schema.GroupVersionKind
+	TargetGVK schema.GroupVersionKind
+	Raw       RawObject
+}
+
+// RawObject represents the raw bytes of the object and its encoding, optionally with a decoded version of the object,
+// which may be any valid resource.Object implementation.
+type RawObject struct {
+	Raw      []byte                `json:",inline"`
+	Object   resource.Object       `json:"-"`
+	Encoding resource.KindEncoding `json:"-"`
+}
+
+// SubresourceRequest is a request to a custom subresource
+type SubresourceRequest struct {
+	ResourceIdentifier resource.FullIdentifier
+	SubresourcePath    string
+	Method             string
+	Headers            http.Header
+	Body               []byte
+}
+
+// Config is the app configuration used in a Provider for instantiating a new App.
+// It contains kubernetes configuration for communicating with an API server, the App's ManifestData as fetched
+// by the runner, and additional arbitrary configuration details that may be app-specific.
+type Config struct {
+	// KubeConfig is a kubernetes rest.Config used to communicate with the API server where the App's Kinds are stored.
+	KubeConfig rest.Config
+	// ManifestData is the fetched ManifestData the runner is using for determining app kinds and capabilities.
+	ManifestData ManifestData
+	// ExtraConfig is additional configuration information, which may be app or runner specific
+	ExtraConfig map[string]any
+}
+
+// Provider represents a type which can provide an app manifest, and create a new App when given a configuration.
+// It should be used by runners to determine an app's capabilities and create an instance of the app to run.
+type Provider interface {
+	// Manifest returns a Manifest, which may contain ManifestData or may point to a location where ManifestData can be fetched from.
+	// The runner should use the ManifestData to determine app capabilities.
+	Manifest() Manifest
+	// NewApp creates a new App instance using the provided config, or returns an error if an App cannot be instantiated.
+	NewApp(Config) (App, error)
+}
```

Comment:
SUGGESTION – I think provider should provide both config and manifest, separately. That way we can completely separate the bootstrap phase from the actual running of the app and use different bootstrap configs used by the runtime for fetching manifests & configs from configs for the apps.

```suggestion
// Config ...
type Config struct {
	// KubeConfig is a kubernetes rest.Config used to communicate with the API server where the App's Kinds are stored.
	KubeConfig rest.Config
	// ExtraConfig is additional configuration information, which may be app or runner specific
	ExtraConfig map[string]any
}

// Provider represents a type which can provide an app manifest, and create a new App when given a configuration.
// It should be used by runners to determine an app's capabilities and create an instance of the app to run.
type Provider interface {
	// Manifest returns a Manifest, which may contain ManifestData or may point to a location where ManifestData can be fetched from.
	// The runner should use the ManifestData to determine app capabilities.
	Manifest() Manifest

	// Config ...
	Config() Config

	// NewApp creates a new App instance using the provided config, or returns an error if an App cannot be instantiated.
	NewApp(Manifest, Config) (App, error)
}
```

Reply:
I think that ideally the runner would be responsible for fetching & loading the app config from various sources (env, args, remote API) and passing it to the app.

It should definitely consist of at least two parts – general config that is known to the runner (e.g. kubeconfig, ports for listeners, API endpoints, etc. etc.) and app-specific config, opaque to the runner.

I suggest that the former should be codified here in the Provider, the latter I'm less certain about but IMHO we _could_ start by having some provider-opaque config (which could e.g. be `resource.Object` or `map[string]any`), which is part of the composite config. The app then would be responsible for properly parsing / casting the config data, maybe that parsing / casting code could be generated via codegen in the future.

Reply:
I've introduced `app.SpecificConfig` (which right now is just an alias to `any`), which is now included in `app.Config` in place of `ExtraConfig`. `app.Provider` now also requires `SpecificConfig()` which returns the app-specific `SpecificConfig`.